Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -200,15 +353,11 @@ - (NSOperationQueue*)_doMainWork {
         didWork = FALSE;
 
         for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-            [priv->suspendedCondition lock];
-            while (priv->isSuspended) {
-                [priv->suspendedCondition wait];
-            }
-            [priv->suspendedCondition unlock];
-
-            if (RunOperationFromLists(&priv->myQueues[i], &priv->queues[i], &priv->curOperation)) {
-                didWork = TRUE;
-            }
+            // Wait until the operation queue is no longer suspended.
+            // NOTE: if an operation queue becomes suspended after this point, the "in flight" operation
+            // will not be suspended, it will continue as if the queue is still running.
+            _priv->WaitUntilNotSuspended();
+            didWork |= _priv->RunOperationFromLists(i);
```

Comment:
short circuit here would not be advisable. The logic here is pretty dastardly in that it for loops without a didWork check in the while loop. This means that even in the cases where work already occured, we should be trying the other priorities. It also means we can't just set it because we don't want to invalidate a previous priority that did in fact do work.

---
In reply to: [95457228](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95457228) [](ancestors = 95457228,95450897)